# Week 3 Activity: Reflection on a Hiring Filter Algorithm 
Activity adapted from exercise developed by [Evan Peck](https://evanpeck.github.io/)

## Scenario: Moogle’s Hiring Filter
Imagine you are working for Moogle, a well-known tech company that receives tens of thousands of job applications from graduating seniors every year. Since the company receives too many job applications for HR to individually assess in a reasonable amount of time, you are asked to create a program that algorithmically analyzes applications and selects the ones most worth passing onto HR.



### Applicant Data
It’s difficult to create these first-pass cuts, so Moogle designs their application forms to get some numerical data about their applicants’ education. Job applicants must enter the grades they received in 6 core CS courses, as well as their overall GPA. For your convenience, this will be stored in a python `list` that you can access. 

For example, a student who received the following scores. . .
- Intro to CS: 100
- Data Structures: 95
- Software Engineering: 80
- Algorithms: 89
- Computer Organization: 91
- Operative Systems: 75
- Overall GPA: 83

. . . would result in the following list: `[100, 95, 80, 89, 91, 75, 83]`. You can assume that index `0` is always Intro to CS, `1` is always Data Structures, and so on.

Because you are processing many applications, your program will receive a list of lists. For example, this would be the information for 3 applicants:

`[ [100, 95, 80, 89, 91, 75, 83], [75, 80, 85, 90, 85, 88, 90], [85, 70, 99, 100, 81, 82, 91] ]`

### Your Task
Your job is to:
1. Determine how you are going to select the top applicants to pass onto HR.
2. Given a list of applicant data (a list of lists), write code to identify a new list of worthwhile candidates.

### The Data 

Before we use the entire dataset of applications, we're going to write and test our code using a much smaller sample of the dataset. This will be saved in `sample_data` and contain only ten applicant lists. Notice how this is just a list of lists with each list being a unique applicant. 

In [1]:
sample_data = [[93, 89, 63, 88, 60, 73, 80], [100, 63, 57, 96, 58, 71, 78], [81, 91, 99, 78, 57, 87, 86], [81, 73, 100, 57, 91, 60, 66], [86, 89, 64, 81, 69, 93, 92], [78, 63, 88, 95, 59, 98, 90], [55, 74, 68, 55, 69, 94, 80], [64, 77, 75, 92, 77, 72, 83], [95, 58, 92, 62, 77, 64, 59], [94, 78, 84, 83, 68, 63, 76]]

In [2]:
len(sample_data)

10

Now let's take a look at the entire dataset of 5,000 applications. This is saved in another file so we're going to go ahead and load this into memory and then take a look at the data. Notice that it's formatted exactly the same as the sample data as a list of lists!

In [3]:
from applications import *

In [4]:
applications

[[93, 89, 63, 88, 60, 73, 80],
 [100, 63, 57, 96, 58, 71, 78],
 [81, 91, 99, 78, 57, 87, 86],
 [81, 73, 100, 57, 91, 60, 66],
 [86, 89, 64, 81, 69, 93, 92],
 [78, 63, 88, 95, 59, 98, 90],
 [55, 74, 68, 55, 69, 94, 80],
 [64, 77, 75, 92, 77, 72, 83],
 [95, 58, 92, 62, 77, 64, 59],
 [94, 78, 84, 83, 68, 63, 76],
 [82, 96, 79, 89, 87, 93, 61],
 [63, 92, 79, 86, 58, 79, 69],
 [87, 73, 62, 59, 77, 94, 82],
 [92, 60, 81, 85, 61, 58, 81],
 [99, 66, 98, 60, 96, 80, 91],
 [56, 76, 76, 88, 73, 72, 91],
 [77, 75, 83, 60, 95, 75, 95],
 [55, 90, 80, 90, 78, 99, 70],
 [59, 57, 61, 69, 93, 88, 96],
 [80, 76, 91, 71, 89, 78, 59],
 [96, 66, 91, 95, 55, 77, 90],
 [68, 77, 70, 79, 59, 88, 97],
 [93, 78, 78, 71, 58, 92, 72],
 [84, 71, 69, 99, 63, 100, 67],
 [79, 81, 74, 91, 66, 89, 62],
 [80, 88, 80, 60, 81, 72, 66],
 [70, 63, 57, 88, 81, 61, 92],
 [81, 100, 86, 97, 72, 71, 58],
 [79, 99, 70, 72, 76, 66, 70],
 [94, 97, 68, 55, 88, 91, 70],
 [66, 89, 97, 66, 90, 71, 100],
 [61, 55, 76, 56, 59, 73, 71],
 [5

In [5]:
len(applications)

5000

### Algorithms 

Now we're going to write algorithms to select applicants based on a variety of decisions. For each one, take note of how many applicants are passed onto the next stage of the application process. Think through the decisions being made and how these decisions might enforce or reduce systemic and cultural oppression. 

1. Selects applicants that have an overall GPA above 80
2. Selects applicants that have no grade below 65
3. Selects applicants that have at least 5 grades above 80
4. Selects applicants that have an average of the six classes above 85
5. Your own algorithm to select applicants 

In [6]:
selection = [] ##1. create a list to hold the selected applicants

for applicant in applications:
    if applicant[6] > 80:
        selection.append(applicant)
        #print(selection)

print(selection)


[[81, 91, 99, 78, 57, 87, 86], [86, 89, 64, 81, 69, 93, 92], [78, 63, 88, 95, 59, 98, 90], [64, 77, 75, 92, 77, 72, 83], [87, 73, 62, 59, 77, 94, 82], [92, 60, 81, 85, 61, 58, 81], [99, 66, 98, 60, 96, 80, 91], [56, 76, 76, 88, 73, 72, 91], [77, 75, 83, 60, 95, 75, 95], [59, 57, 61, 69, 93, 88, 96], [96, 66, 91, 95, 55, 77, 90], [68, 77, 70, 79, 59, 88, 97], [70, 63, 57, 88, 81, 61, 92], [66, 89, 97, 66, 90, 71, 100], [55, 100, 77, 93, 63, 73, 81], [73, 68, 97, 65, 91, 69, 92], [70, 98, 69, 81, 84, 70, 98], [92, 58, 97, 83, 84, 70, 95], [87, 93, 63, 60, 81, 65, 84], [90, 84, 55, 81, 80, 89, 93], [62, 82, 67, 88, 69, 58, 89], [87, 100, 93, 82, 87, 81, 83], [62, 59, 75, 65, 72, 62, 93], [68, 82, 71, 93, 72, 60, 97], [90, 99, 61, 63, 83, 67, 86], [97, 78, 81, 92, 61, 91, 84], [60, 84, 73, 90, 96, 73, 90], [85, 85, 93, 77, 100, 59, 100], [75, 99, 67, 79, 86, 93, 95], [71, 83, 64, 72, 70, 63, 90], [89, 93, 84, 59, 58, 69, 94], [69, 82, 83, 77, 67, 64, 84], [98, 56, 77, 69, 94, 67, 84], [96,

How many applicants did we select with this filter? What percentage of our total sample size is that?  and 2117

Let's try another potential algorithm where we select applicants that have no grade below a 65. 

In [7]:
print(len(selection)) ##How many applicants did we select with this filter? What percentage of our total sample size is that?  


print(len(selection)/len(applications)*100) ##hat percentage of our total sample size is that?  


2117
42.34


In [8]:
selection_nograde_below65 = [] ## create a list to hold the selected applicants
for applicant in applications:
    for grade in applicant:
        if grade >= 65:
            selection_nograde_below65.append(applicant)
            ## print(applicant)


In [9]:
len(selection_nograde_below65)

27464

Now let's try to create a filter for selecting applicants that have at least five grades above 85. **Hint**: A counter can be useful here! 

In [9]:

selection_atleast_5gradesabove85 = [] ## create a list to hold the selected applicants with five grades above 85

for applicant in applications:
    count = 0
    for grade in applicant:
        if grade > 85:
            count+=1
        if count >= 5:
            selection_atleast_5gradesabove85.append(applicant)
            #print(selection_atleast_5gradesabove85)

print(selection_atleast_5gradesabove85)

[[100, 95, 61, 99, 91, 98, 75], [100, 95, 61, 99, 91, 98, 75], [88, 70, 99, 92, 67, 96, 89], [91, 57, 88, 96, 90, 90, 80], [91, 57, 88, 96, 90, 90, 80], [93, 99, 92, 91, 88, 91, 100], [93, 99, 92, 91, 88, 91, 100], [93, 99, 92, 91, 88, 91, 100], [96, 91, 85, 91, 92, 80, 93], [72, 99, 95, 91, 55, 98, 90], [94, 64, 88, 91, 83, 89, 95], [57, 62, 91, 93, 90, 86, 95], [98, 94, 95, 95, 76, 99, 90], [98, 94, 95, 95, 76, 99, 90], [65, 89, 78, 93, 98, 99, 98], [98, 84, 98, 77, 91, 94, 94], [55, 93, 84, 98, 89, 88, 91], [91, 90, 96, 100, 56, 98, 71], [91, 90, 96, 100, 56, 98, 71], [88, 88, 88, 72, 95, 73, 99], [95, 90, 96, 68, 73, 90, 89], [88, 88, 94, 77, 61, 89, 92], [98, 95, 96, 68, 70, 95, 98], [87, 99, 87, 80, 89, 95, 61], [87, 99, 87, 80, 89, 95, 61], [58, 95, 90, 84, 86, 89, 91], [72, 88, 92, 86, 92, 64, 93], [70, 100, 88, 69, 94, 87, 95], [98, 96, 92, 73, 86, 92, 93], [98, 96, 92, 73, 86, 92, 93], [80, 55, 97, 100, 90, 87, 90], [96, 86, 94, 97, 65, 92, 56], [96, 86, 94, 97, 65, 92, 56], 

In [10]:
len(selection_atleast_5gradesabove85)

298

Finally, let's write a algorithm that selects applicants if they have an average grade of at least 85 across their six classes. 

In [11]:
selection_avggradesabove85 = [] ## create a list to hold the selected applicants with five grades above 85


for applicant in applications:
    total = 0  # Reset sum for each applicant
    for grade in applicant :
        total += grade  # Sum up grades

    if total / 7 >= 85:  # Check if the average is at least 85
        selection_avggradesabove85.append(applicant)  # Append to the list

print(selection_avggradesabove85)


[[87, 100, 93, 82, 87, 81, 83], [85, 85, 93, 77, 100, 59, 100], [96, 85, 61, 81, 87, 90, 98], [93, 91, 83, 75, 94, 81, 89], [74, 61, 100, 91, 84, 89, 96], [100, 95, 61, 99, 91, 98, 75], [95, 99, 85, 98, 59, 93, 73], [92, 85, 90, 99, 87, 85, 79], [88, 70, 99, 92, 67, 96, 89], [69, 79, 77, 94, 94, 100, 98], [71, 97, 76, 84, 100, 88, 91], [93, 99, 92, 91, 88, 91, 100], [96, 91, 85, 91, 92, 80, 93], [72, 99, 95, 91, 55, 98, 90], [94, 64, 88, 91, 83, 89, 95], [99, 67, 97, 67, 76, 98, 96], [62, 85, 99, 92, 84, 84, 93], [65, 91, 87, 99, 100, 79, 76], [68, 74, 97, 76, 95, 99, 93], [98, 94, 95, 95, 76, 99, 90], [65, 89, 78, 93, 98, 99, 98], [98, 84, 98, 77, 91, 94, 94], [55, 93, 84, 98, 89, 88, 91], [98, 97, 82, 68, 95, 85, 78], [100, 98, 99, 60, 97, 75, 70], [100, 75, 92, 81, 79, 99, 83], [91, 90, 96, 100, 56, 98, 71], [88, 88, 88, 72, 95, 73, 99], [72, 84, 100, 96, 74, 79, 96], [100, 73, 74, 100, 90, 90, 72], [95, 90, 96, 68, 73, 90, 89], [78, 61, 99, 97, 74, 100, 89], [77, 83, 88, 91, 95, 81

In [12]:
print(len(selection_avggradesabove85))

361


In the space below, work in group to decide what types of criteria you want to use to write an algorithm to select applicants. First test it with the sample data then run it with the entire set of applicants.  

A useful piece of code that will give you the percentage of applicants you kept is: 

`print("Your algorithm kept", round(len(selection)/len(applications)*100), "percent of applicants")`

In [15]:
## your algorithm 
selection_avggradesabovegpa = [] 

for applicant in applications:
    total = 0  # Reset sum for each applicant
    for grade in applicant :
        total += grade  # Sum up grades

    if total / 7 >= applicant[6]:  # Check if the average is at least 85
        selection_avggradesabovegpa.append(applicant)  # Append to the list

print(selection_avggradesabovegpa)
print(len(selection_avggradesabovegpa))

print("Your algorithm kept", round(len(selection)/len(applications)*100), "percent of applicants")


[[81, 73, 100, 57, 91, 60, 66], [95, 58, 92, 62, 77, 64, 59], [94, 78, 84, 83, 68, 63, 76], [82, 96, 79, 89, 87, 93, 61], [63, 92, 79, 86, 58, 79, 69], [55, 90, 80, 90, 78, 99, 70], [80, 76, 91, 71, 89, 78, 59], [93, 78, 78, 71, 58, 92, 72], [84, 71, 69, 99, 63, 100, 67], [79, 81, 74, 91, 66, 89, 62], [80, 88, 80, 60, 81, 72, 66], [81, 100, 86, 97, 72, 71, 58], [79, 99, 70, 72, 76, 66, 70], [94, 97, 68, 55, 88, 91, 70], [94, 78, 70, 88, 76, 93, 70], [72, 77, 91, 97, 59, 89, 68], [97, 88, 60, 63, 75, 94, 58], [79, 64, 88, 96, 55, 82, 65], [82, 76, 61, 70, 84, 64, 70], [97, 71, 59, 67, 72, 57, 56], [88, 58, 93, 98, 58, 76, 60], [60, 80, 97, 64, 98, 79, 74], [87, 100, 93, 82, 87, 81, 83], [78, 89, 85, 55, 72, 85, 75], [80, 88, 88, 79, 59, 88, 58], [70, 70, 89, 87, 99, 60, 69], [69, 96, 70, 90, 91, 91, 70], [93, 58, 97, 85, 92, 85, 79], [65, 92, 71, 86, 73, 57, 56], [77, 55, 67, 72, 75, 74, 56], [88, 62, 67, 86, 55, 60, 69], [92, 60, 84, 100, 94, 63, 60], [87, 67, 92, 88, 98, 76, 76], [99,

Questions to Answer: 
1. What criteria did you choose to select finalists? How did you choose that criteria?
2. Roughly what percentage of applicants does your algorithm pass on as finalists? Is that enough? If Moogle asked you to take a more aggressive approach with your algorithm, are there any tradeoffs?

1. I chose to select applicants based on their average grade. The criteria I used to select the finalists was to find applicants whose average grade is above their gpa. This is because I thought for some of the applicants that they might have encountered semesters where they didn't do well and it afected their GPA. This also accounts for courses that might not have been listed but probably affected their GPA. 

2. My algorithm kept 42% of the applicants. I think this is enought because if kept at most less than 50% of the applicants. This is a good quantity for first pass applicants as it gives Moogle a large enough pool to begin the second pass stage. If a more aggressive approach is used, it might let them loose quality applicants who might not necessary meet all the stringent measures thata re used for the first pass phase. 



___
While our data seemed to perfectly reported and without any inconsitencies, the world is less perfect. Consider the following scenarios: 

___
*Story 1*: Misread the Instructions
What if an excellent applicant thinks they should put in letter grades?

`[‘A’, ‘A’, ‘A’, ‘A’, ‘A’, ‘A’, ‘A’]`

. . . or how about their grades on 4-point scale?

`[4, 3.9, 4, 4, 3.95, 4, 3.9]`

___
*Story 2*: Bad Assumptions
What if one of your applicants skipped Intro to Computer Science? When they saw your form, they froze, and decided that putting -1 in the input field would make it obvious. . .

`[-1, 95, 99, 94, 96, 98, 95]`

___
*Story 3*: Mistake in the Input
What if one of your applicants accidentally put in a number > 100?

`[681, 68, 73, 70, 81, 91, 59]`

That might seem easy enough for a program to catch, but what if they accidentally dropped a 0?

`[100, 100, 100, 100, 100, 100, 10]`

A person would catch that mistake easily, does your algorithm?

___
*Story 4*: The Awful Semester
What if your applicant had a medical emergency one semester? Or a personal tragedy?

`[95, 93, 50, 91, 98, 90, 90]`

___
*Story 5*: Inverse Trajectories
What if one of your applicants came from an underprivileged background and really struggled at the beginning of college. . . but showed extraordinary growth by the end?

`[65, 75, 85, 95, 100, 100, 80]`

What if one of your applicants came to college with extraordinary potential? They easily aced their first few classes and then gradually grew apathetic about their education - getting nothing but barely-passing grades by the time they were a senior?

`[100, 100, 95, 85, 75, 65, 80]`

Does your algorithm treat them equally?

___

Complete the following questions reflecting on these scenarios:

3. What systemic advantages/disadvantages are your algorithms likely to amplify?

4. What does it mean to design a fair algorithm?

5. If you had access to additional data beyond grades (e.g., extracurricular activities, internships, letters of recommendation), how might you incorporate it into your selection process? Would it make your algorithm fairer or introduce new biases?

6. How do current hiring filter algorithms work? What problems do they encounter? How do these algorithms broadly compare to the ones we wrote today? (Some example articles discussing hiring filters linked below in citations but you're not limited to these examples) 


3. This algorithm amplifies the advantages of students from privileged backgrounds. Those with consistent access to resources, stable family lives, and quality education are more likely to have consistently high grades. This algorithm disadvantages students facing systemic challenges like poverty, discrimination, or personal hardship. These students may have inconsistent grades due to factors outside their control, even if they possess great potential. In Story 4, The Awful Semester and Story 5,  Inverse Trajectories, a single bad semester or an upward trajectory of improvement isn't adequately considered. This further marginalizes students who may have overcome significant obstacles to achieve academic success. Furthermore, the algorithm's rigidity, as seen in Story 1, Misread the Instructions, penalizes those unfamiliar with grading conventions in the US if they are international applicants or any local student who misread it.

4.To design a fair algorithm means to have an algorithm that minimizes bias and provides an equal grounds for all applicants to be assessed.There should be an understanding of the limitations of using grades as the sole measure of potential. Grades are a product of a complex system and do not fully capture an individual's abilities or potential. Potential biases in the data itself, like the historical inequities influencing access to quality education must be addressed
The algorithm's decision-making process should also be made transparent and understandable to applicants. This allows for scrutiny and identification of potential biases.


5. Including additional data like extracurriculars, internships, and recommendations can provide a more holistic view of the applicant. However, this can also introduce new biases. Students from privileged backgrounds may have greater access to prestigious extracurricular activities and internships, further reinforcing existing advantages.
Letters of recommendation can be biased, consciously or unconsciously, reflecting societal prejudices based on race, gender, or socioeconomic status.
To mitigate these biases, we can introduce a Standardized Evaluation. Using standardized rubrics to evaluate qualitative data like extracurriculars and recommendations, will help to minimize subjective interpretations. Also, we can incorporate a qualitative review process where humans consider the context and nuances of each application, especially in cases where the algorithm flags potential discrepancies or exceptional circumstances.


6. Current hiring filter algorithms, often used by large companies, analyze resumes and applications for keywords, skills, and experience. They encounter problems such as:
Bias Amplification. They akso often fail to capture the nuances of human experience and potential, overlooking non-traditional career paths or skills gained through unconventional means.Applicants can also learn to "game" the system by putting on their resumes keywords, even if they lack genuine expertise in those areas.
These algorithms share similarities with the admissions algorithm discussed in that they rely on quantifiable data and can lead to systemic inequalities. However, they often incorporate more diverse data points, including skills and work history. The key takeaway is that any algorithm used in high-stakes decision-making must be carefully designed and evaluated for fairness and potential biases limited as much as possible. 

Citations: 

Stacy A. Doore, Casey Fiesler, Michael S. Kirkpatrick, Evan Peck, and Mehran Sahami. 2020. Assignments that Blend Ethics and Technology. In Proceedings of the 51st ACM Technical Symposium on Computer Science Education (SIGCSE ‘20). Association for Computing Machinery, New York, NY, USA, 475–476. DOI:https://doi.org/10.1145/3328778.3366994

[Amazon scraps secret AI recruiting tool that showed bias against women](https://www.reuters.com/article/us-amazon-com-jobs-automation-insight/amazon-scraps-secret-ai-recruiting-tool-that-showed-bias-against-women-idUSKCN1MK08G/)

[Hiring Algorithms Are Not Neutral](https://hbr.org/2016/12/hiring-algorithms-are-not-neutral)

[Can an Algorithm Hire Better Than a Human?](https://www.nytimes.com/2015/06/26/upshot/can-an-algorithm-hire-better-than-a-human.html)

[Algorithms in Hiring](https://blog.learningcollider.org/algorithms-in-hiring-6760ea8869b)

[Exploration-based algorithms can improve hiring quality and diversity](https://mitsloan.mit.edu/ideas-made-to-matter/exploration-based-algorithms-can-improve-hiring-quality-and-diversity)

[AI hiring tools may be filtering out the best job applicants](https://www.bbc.com/worklife/article/20240214-ai-recruiting-hiring-software-bias-discrimination)

[Challenges for mitigating bias in algorithmic hiring](https://www.brookings.edu/articles/challenges-for-mitigating-bias-in-algorithmic-hiring/)

